File: tests/k.go
package main
import ("os";"fmt")
func rot(n,x,y,rx,ry int) (int,int) { if ry==0 { if rx==1 { x,y=n-1-x,n-1-y }; return y,x }; return x,y }
func d2xy(n,d int) (int,int) { x,y,t:=0,0,d; for s:=1;s<n;s*=2 { rx:=1&(t/2); ry:=1&(t^rx); x,y=rot(s,x,y,rx,ry); x+=s*rx; y+=s*ry; t/=4 }; return x,y }
func main() {
    p:="vines_images/go_vine.ppm"; f,_:=os.Create(p); f.Write([]byte("P6\n4000 4000\n255\n")); f.Truncate(48000017); f.Close()
    f,_=os.OpenFile(p,os.O_RDWR,0644); h:=543210
    for i:=121;i<=160;i++ {
        s:=fmt.Sprintf("SQL_ID_%d_DATA\x00",i); cur:=h+(i*1000)
        for j:=0;j<len(s);j+=3 {
            x,y:=d2xy(4096,cur); end:=j+3; if end>len(s){end=len(s)}; c:=make([]byte,3); copy(c, s[j:end])
            f.WriteAt(c,int64(17+(y*4000+x)*3)); cur++
        }
    }
}

File: tests/k.cpp
#include <iostream>
#include <fstream>
#include <string>
void rot(int n,int* x,int* y,int rx,int ry){ if(ry==0){ if(rx==1){ *x=n-1-*x; *y=n-1-*y; } int t=*x; *x=*y; *y=t; } }
void d2xy(int n,int d,int& x,int& y){ int rx,ry,s,t=d; x=y=0; for(s=1;s<n;s*=2){ rx=1&(t/2); ry=1&(t^rx); rot(s,&x,&y,rx,ry); x+=s*rx; y+=s*ry; t/=4; } }
int main(){
    std::string p="vines_images/cpp_vine.ppm"; std::ofstream o(p,std::ios::binary); o<<"P6\n4000 4000\n255\n"; o.seekp(48000016); o.put(0); o.close();
    std::fstream f(p,std::ios::in|std::ios::out|std::ios::binary); int h=543210;
    for(int i=81;i<=120;i++){
        std::string s="SQL_ID_"+std::to_string(i)+"_DATA"; s+='\0'; int cur=h+(i*1000);
        for(size_t j=0;j<s.length();j+=3){
            int x,y; d2xy(4096,cur,x,y); f.seekg(17+(y*4000+x)*3);
            std::string c=s.substr(j,3); while(c.length()<3) c+='\0';
            f.write(c.c_str(),3); cur++;
        }
    } return 0;
}
